File: Recursion/subsequence_pattern/GenerateParenthesis.cpp
#include <string>
#include <vector>

class GenerateParenthesis {
    static void helper(int n, std::vector<std::string>& result, const std::string& current,
                       int open, int close) {
        if (static_cast<int>(current.size()) == 2 * n) { result.push_back(current); return; }
        if (open < n) helper(n, result, current + '(', open + 1, close);
        if (close < open) helper(n, result, current + ')', open, close + 1);
    }
public:
    static std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> result;
        helper(n, result, "", 0, 0);
        return result;
    }
};

File: Recursion/subsequence_pattern/GenerateParenthesis.py
def _helper(pairs: int, result: list[str], current: str, opened: int, closed: int) -> None:
    if len(current) == 2 * pairs:
        result.append(current)
        return
    if opened < pairs:
        _helper(pairs, result, current + "(", opened + 1, closed)
    if closed < opened:
        _helper(pairs, result, current + ")", opened, closed + 1)


def generate_parenthesis(pairs: int) -> list[str]:
    result = []
    _helper(pairs, result, "", 0, 0)
    return result
